File: NHSOnline.Backend.Worker.IntegrationTests/Worker/WorkerClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using NHSOnline.Backend.Worker.IntegrationTests.Worker.Models.Patient;
using NHSOnline.Backend.Worker.IntegrationTests.Worker.Models.User;

namespace NHSOnline.Backend.Worker.IntegrationTests.Worker
{
    public class WorkerClient
    {
        private static readonly JsonSerializerSettings _jsonSerializerSettings = new JsonSerializerSettings
        {
            ContractResolver = new CamelCasePropertyNamesContractResolver()
        };

        private readonly HttpClient _client;

        public WorkerClient()
        {
            _client = new HttpClient
            {
                BaseAddress = new Uri(Configuration.BackendBaseUrl)
            };
        }

        public async Task<Im1ConnectionResponse> GetIm1Connection(string connectionToken, string odsCode)
        {
            var request = new HttpRequestMessage(HttpMethod.Get, WorkerPaths.PatientIm1Connection);
            request.Headers.Add(WorkerHeaders.ConnectionToken, connectionToken);
            request.Headers.Add(WorkerHeaders.OdsCode, odsCode);
            var response = await SendAsync(request);
            var json = await response.Content.ReadAsStringAsync();
            var res = JsonConvert.DeserializeObject<Im1ConnectionResponse>(json, _jsonSerializerSettings);
            return res;
        }

        public async Task<Im1ConnectionResponse> PostIm1Connection(Im1ConnectionRequest requestBody)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, WorkerPaths.PatientIm1Connection)
            {
                Content = new StringContent(JsonConvert.SerializeObject(requestBody), Encoding.UTF8, "application/json")

            };

            var response = await SendAsync(request);
            var json = await response.Content.ReadAsStringAsync();

            return JsonConvert.DeserializeObject<Im1ConnectionResponse>(json, _jsonSerializerSettings);
        }

        public async Task<SessionResponseTestObject> PostSession(UserSessionRequest requestBody)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, WorkerPaths.PatientSessionConnection)
            {
                Content = new StringContent(JsonConvert.SerializeObject(requestBody), Encoding.UTF8, "application/json")
            };

            var response = await SendAsync(request);

            var json = await response.Content.ReadAsStringAsync();

            var sessionResponseTestObject = new SessionResponseTestObject
            {
                UserSessionResponse = JsonConvert.DeserializeObject<UserSessionResponse>(json, WorkerClient._jsonSerializerSettings)
            };

            if (response.Headers.TryGetValues("Set-Cookie", out var values))
            {
                sessionResponseTestObject.Cookie = values.First();
            }

            return sessionResponseTestObject;
        }


        private async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request)
        {
            var response = await _client.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                // Exception is thrown here to ensure that the tests fail at the appropriate location and not further down the line
                // when values are not as expected.  This makes it easier to debug.
                throw new NhsoHttpException(request, response);
            }

            return response;
        }
    }
}

File: src/main/kotlin/worker/WorkerClient.kt
package worker

import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import config.Config

import org.apache.http.HttpResponse
import org.apache.http.HttpStatus.SC_CREATED
import org.apache.http.HttpStatus.SC_OK
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpPost
import org.apache.http.client.methods.HttpUriRequest
import org.apache.http.impl.client.HttpClients
import worker.models.patient.Im1ConnectionRequest
import worker.models.patient.Im1ConnectionResponse
import org.apache.http.entity.StringEntity
import org.apache.http.protocol.HttpContext
import worker.models.courses.CourseListResponse
import worker.models.prescriptions.PrescriptionListResponse
import worker.models.session.UserSessionRequest
import worker.models.session.UserSessionResponse
import worker.models.session.UserSessionResponse.UserSessionResponseCookie
import java.io.InputStreamReader
import java.io.BufferedReader
import java.net.URLEncoder
import javax.servlet.http.Cookie


class WorkerClient {

    var cookieHeaderKey = "Set-Cookie"
    private val _client: HttpClient
    private val gsonBuilder: GsonBuilder = GsonBuilder()
    private var gson: Gson
    private val config = Config.instance

    init {
        _client = HttpClients.createDefault()
        gsonBuilder.setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
        gson = gsonBuilder.create()
    }


    fun getIm1Connection(connectionToken: String?, odsCode: String?): Im1ConnectionResponse {
        val httpGet = HttpGet(config.backendUrl + WorkerPaths.patientIm1Connection)
        httpGet.setHeader(WorkerHeaders.ConnectionToken, connectionToken)
        httpGet.setHeader(WorkerHeaders.OdsCode, odsCode)

        val response = sendAsync(httpGet)
        val rd = BufferedReader(InputStreamReader(response.entity.content))
        val result = rd.use { it.readText() }
        httpGet.releaseConnection()

        return gson.fromJson(result, Im1ConnectionResponse::class.java)

    }

    fun postIm1Connection(requestBody: Im1ConnectionRequest): Im1ConnectionResponse {
        val httpPost = HttpPost(config.backendUrl + WorkerPaths.patientIm1Connection)
        val entity = StringEntity(gson.toJson(requestBody), "UTF-8")
        entity.setContentType("application/json")
        httpPost.entity = entity

        val response = sendAsync(httpPost)
        val rd = BufferedReader(InputStreamReader(response.entity.content))
        val result = rd.use { it.readText() }
        httpPost.releaseConnection()
        println(result)
        return gson.fromJson<Im1ConnectionResponse>(result, Im1ConnectionResponse::class.java)
    }

    fun postSessionConnection(requestBody: UserSessionRequest): UserSessionResponse {
        val httpPost = HttpPost(config.backendUrl + WorkerPaths.sessionConnection)
        val entity = StringEntity(gson.toJson(requestBody), "UTF-8")
        entity.setContentType("application/json")
        httpPost.entity = entity

        val response = sendAsync(httpPost)
        val rd = BufferedReader(InputStreamReader(response.entity.content))
        val result = rd.use { it.readText() }
        httpPost.releaseConnection()
        println(result)

        val userSessionResponseBody = gson.fromJson<UserSessionResponse.UserSessionResponseBody>(result,
                UserSessionResponse.UserSessionResponseBody::class.java)

        val userSessionResponseCookie = UserSessionResponseCookie(Cookie(cookieHeaderKey, response.getHeaders(cookieHeaderKey).first().value))
        return UserSessionResponse(userSessionResponseCookie, userSessionResponseBody)
    }

    fun getPrescriptionsConnection(fromDate: String?, context: HttpContext?): PrescriptionListResponse {
        var queryString = ""
        if (fromDate != null) queryString = "?FromDate=" + URLEncoder.encode(fromDate, "UTF-8")
        val httpGet = HttpGet(config.backendUrl + WorkerPaths.getPrescriptionsConnection + queryString)
        val response = sendAsync(httpGet, context)
        val rd = BufferedReader(InputStreamReader(response.entity.content))
        val result = rd.use { it.readText() }
        httpGet.releaseConnection()

        return gson.fromJson<PrescriptionListResponse>(result, PrescriptionListResponse::class.java)
    }


    fun getCoursesConnection(context: HttpContext?): CourseListResponse {
        val httpGet = HttpGet(config.backendUrl + WorkerPaths.getCoursesConnection)

        val response = sendAsync(httpGet, context)
        val rd = BufferedReader(InputStreamReader(response.entity.content))
        val result = rd.use { it.readText() }
        httpGet.releaseConnection()

        return gson.fromJson<CourseListResponse>(result, CourseListResponse::class.java)
    }


    private fun sendAsync(request: HttpUriRequest, context: HttpContext? = null): HttpResponse {
        val response = if (context != null) _client.execute(request, context) else _client.execute(request)

        if (response.statusLine.statusCode != SC_OK && response.statusLine.statusCode != SC_CREATED) {
            // Exception is thrown here to ensure that the tests fail at the appropriate location and not further down the line
            // when values are not as expected.  This makes it easier to debug.
            throw NhsoHttpException(request, response)
        } else {
            return response
        }
    }
}